### PR TITLE
fixing bugs in seding mail,featured prodcuts and adding address.

### PR DESCRIPTION
**Pull request by Gangotri-sg3**

### CODE REVIEW

File: client/src/components/Address/EditAddress.js [hunk 1]
```diff
@@ -11,12 +13,16 @@ const EditAddress = () => {
...
   let [pin_code, setPin_code] = useState(null);
   let [state, setState] = useState(null);
-  let [country_id, setCountry_id] = useState(null);
+  let [country, setCountry] = useState(null);
   let [user_id, setUser_id] = useState(1);
-  let [countries, setCountries] = useState(null);
   const [message, setMessage] = useState(null);
   const navigate = useNavigate();
   let id = window.location.pathname.substring(14, 15);
+  const options = useMemo(() => countryList().getData(), []);
```
Comment (iMagesh) [inline client/src/components/Address/EditAddress.js:21]: why are yo using useMemo here?

File: client/src/components/Address/NewAddress.js [hunk 2]
```diff
@@ -1,32 +1,38 @@
 import React from "react";
```
Comment (iMagesh) [inline client/src/components/Address/NewAddress.js:1]: NewAddress and EditAddress component seems to be doing the exact same, the codes are duplicated. Can we have one shared component for both Add/Edit and remove all the duplicates?

File: server/migrations/20220118165920_created_intial_tables.js [hunk 3]
```diff
@@ -1,162 +1,165 @@
...
-    .createTable("cart_items", (table) => {
-      table.increments("id").primary();
-      table.integer("quantity").notNullable();
-      table.integer("variant_id").unsigned().notNullable();
-      table.foreign("variant_id").references("variants.id");
-      table.integer("cart_id").unsigned().notNullable();
-      table.foreign("cart_id").references("cart.id");
-      table.timestamps(true, true);
-    });
+  return (
```
Comment (iMagesh) [inline server/migrations/20220118165920_created_intial_tables.js:2]: why is this file showing as updated? what did you change here? make sure you don't remove any of the changes that others might have done.

File: server/package.json [hunk 4]
```diff
@@ -20,10 +20,12 @@
     "dotenv": "^14.2.0",
     "express": "^4.17.1",
     "express-validator": "^6.14.0",
+    "form-data": "^4.0.0",
```
Comment (iMagesh) [inline server/package.json:23]: what is this for?

File: server/routes/forgotPassword.js [hunk 5]
```diff
@@ -4,6 +4,15 @@ const knex = require("../utils/dbConfig");
 const jwt = require("jsonwebtoken");
 const transporter = require("../utils/nodemailer");
 
+const API_KEY = process.env.API_KEY;
+const DOMAIN = process.env.DOMAIN_NAME;
+
+const formData = require("form-data");
```
Comment (iMagesh) [inline server/routes/forgotPassword.js:10]: explain why you need this formData here?
I think it is not required, can we use mailgun without this package?

File: server/routes/resetPassword.js [hunk 6]
```diff
@@ -3,43 +3,56 @@ const router = express.Router();
 const knex = require("../utils/dbConfig");
 const bcrypt = require("bcryptjs");
 const transporter = require("../utils/nodemailer");
+const API_KEY = process.env.API_KEY;
+const DOMAIN = process.env.DOMAIN_NAME;
+
+const formData = require("form-data");
+const Mailgun = require("mailgun.js");
+
+const mailgun = new Mailgun(formData);
```
Comment (iMagesh) [inline server/routes/resetPassword.js:12]: mailgun config seems to be duplicated, can we move this to a file and re-use everywhere?

**Comment by iMagesh** [comment]
@Gangotri-sg3 resolve all the above comments and then we can merge

**Comment by iMagesh** [comment]
@Gangotri-sg3 did you push your changes? since you are on leave I can ask someone else to work on it.

File: client/src/components/Address/EditAddress.js (hunk 1 above)
Comment (Gangotri-sg3) [reply client/src/components/Address/EditAddress.js]: The useMemo  is a local call,  immediately renders value.which has only something to do with this component. when  some value is more urgent that time we can use this.
here we need country list soon so we r using usememo ,we cant use useEffect or useState.

The useEffect is a collective call, async or not, it's collected after all components are rendered

File: client/src/components/Address/NewAddress.js (hunk 2 above)
Comment (Gangotri-sg3) [reply client/src/components/Address/NewAddress.js]: fixed

File: server/migrations/20220118165920_created_intial_tables.js (hunk 3 above)
Comment (Gangotri-sg3) [reply server/migrations/20220118165920_created_intial_tables.js:2]: no other changes

File: server/routes/resetPassword.js (hunk 6 above)
Comment (Gangotri-sg3) [reply server/routes/resetPassword.js]: fixed

**Pull request merged**